aspellWithDicts: create derivation with aspell and selected dictionaries

FRidh opened the pull request:
Currently, `aspell` checks the active profiles for dictionaries. While
this may be convenient, it does not work with `nix-shell` and it doesn't
allow any isolation.

This commit adds the possibility to use composition by creating a
derivation with `buildEnv` that contains `aspell` and selected
dictionaries.

Nix example:
```
my_aspell = aspellWithDicts(ps: with ps; [ en nl ])
```
`nix-shell` example:
```
nix-shell -p 'aspellWithDicts(ps: with ps; [ en nl ])'
```

FRidh commented:
Composition always increases space usage. The example derivation is 4 MB. We may choose to create only symbolic links for the dictionaries and then wrappers for the executables.

lheckemann commented on pkgs/development/libraries/aspell/aspell-with-dicts.nix:
```diff
@@ -0,0 +1,31 @@
+  postBuild = ''
+#     wrapProgram $out/bin/aspell --set ASPELL_CONF "data-dir $out/lib/aspell"
+    # We cannot simply use wrapProgram because we're not allowed to modify items in $out/bin
+    unlink $out/bin
```
Why unlink rather than the more conventional rm?

FRidh replied:
ah, of course, removing individual items with `rm` doesn't work because the directory containing them is a symbolic link itself. I actually hadn't tried using `rm` on symbolic link to `/bin`.

lheckemann commented:
Did you consider making the dictionaries directory a separate derivation? That would avoid modifying the result of buildEnv and the need to remove part of what it created. It would also result in the `manifest.nix` remaining accurate.

FRidh commented:
> Did you consider making the dictionaries directory a separate derivation?

Yes, if I understand you correctly. My idea was to have then two derivations, one with wrappers pointing to the other consisting of symlinks to the dicts.

> We may choose to create only symbolic links for the dictionaries and then wrappers for the executables.

bjornfor commented:
I prefer having the option of installing things in my profile. I don't think you need to remove that feature to get what you want. When ASPELL_CONF is set in the pure wrapper, the impure wrapper automatically disables profile traversal. (You can also unset NIX_PROFILES in the pure wrapper just to be sure.)

FRidh commented:
@bjornfor good point, I had missed that. I'll update the PR.

FRidh committed:
aspellWithDicts: create derivation with aspell and selected dictionaries

Currently, `aspell` checks the active profiles for dictionaries. While
this may be convenient, it does not work with `nix-shell` and it doesn't
allow any isolation.

This commit adds the possibility to use composition by creating a
derivation with `buildEnv` that contains `aspell` and selected
dictionaries.

Nix example:

my_aspell = aspellWithDicts(ps: with ps; [ en nl ])

`nix-shell` example:

nix-shell -p 'aspellWithDicts(ps: with ps; [ en nl ])'

FRidh commented:
It may be good if the different spell checkers work similarly.
For `hunspell` there is also a `hunspellWithDicts` which only wrappers `hunspell` https://github.com/NixOS/nixpkgs/blob/release-17.03/pkgs/development/libraries/hunspell/wrapper.nix

alexeymuranov commented:
Just a remark about hunspell.  I didn't manage to make hunspell work even using `hunspellWithDicts`.  Only US English works, and it does so without any special configuration. For some reason, Hebrew shows up in menus without having been (explicitly) installed.

FRidh commented:
> Edit. Maybe i am a bit confused: probably it is not hunspell that does not work, but its dictionaries are not seen by different applications, like Evolution or gEdit.

Applications typically need to be compiled with support and they may assume/hardcode a path where the dictionaries could be. See https://github.com/NixOS/nixpkgs/issues/26654.

FRidh commented:
@lheckemann alternative at https://github.com/FRidh/nixpkgs/commit/3d72a3c0eb883d553d41d3bb5295d15a7d83249f. Note the commit message needs to be updated.

bennofs commented:
@FRidh can you update the PR to the alternate version? I'd really like to see this in nixpkgs :+1:

FRidh commented:
Pushed c0cf19608f to master.
